src/MongoDB.Driver/Encryption/LibMongoCryptController.cs

```diff
@@ -0,0 +1,681 @@
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_NEED_MONGO_KEYS:
+                        ProcessNeedMongoKeysState(context, cancellationToken);
+                        break;
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_NEED_MONGO_MARKINGS:
+                        ThrowIfNotAutoEncryption(context.State);
+                        ProcessNeedMongoMarkingsState(context, databaseName, cancellationToken);
+                        break;
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_READY:
+                        result = ProcessReadyState(context);
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Unexpected context state: {context.State}.");
+                }
+            }
+        }
+
+        private async Task<byte[]> ProcessStatesAsync(CryptContext context, string databaseName, CancellationToken cancellationToken)
+        {
+            byte[] result = null;
+            while (true)
+            {
+                switch (context.State)
+                {
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_DONE:
+                        return result;
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_NEED_KMS:
+                        await ProcessNeedKmsStateAsync(context, cancellationToken).ConfigureAwait(false);
+                        break;
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_NEED_MONGO_COLLINFO:
+                        await ProcessNeedCollectionInfoStateAsync(context, databaseName, cancellationToken).ConfigureAwait(false);
+                        break;
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_NEED_MONGO_KEYS:
+                        await ProcessNeedMongoKeysStateAsync(context, cancellationToken).ConfigureAwait(false);
+                        break;
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_NEED_MONGO_MARKINGS:
+                        ThrowIfNotAutoEncryption(context.State);
+                        await ProcessNeedMongoMarkingsStateAsync(context, databaseName, cancellationToken).ConfigureAwait(false);
+                        break;
+                    case CryptContext.StateCode.MONGOCRYPT_CTX_READY:
+                        result = ProcessReadyState(context); // no async version needed
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Unexpected context state: {context.State}.");
+                }
+            }
+        }
+
+        private void SendKmsRequest(KmsRequest request, CancellationToken cancellation)
+        {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+            socket.Connect(request.Endpoint, 443);
+
+            var validationCallback = new RemoteCertificateValidationCallback(AcceptAnyCertificate); // TODO: what validation needs to be done?
```

Comment:
Hm, should we be using the two callbacks that the user may have provided to us, accessible via the `SslSettings`stored in `MongoClient.MongoClientSettings.SslSettings`?
@rstam